crates/kubelet/src/plugin_watcher/mod.rs

```diff
@@ -0,0 +1,744 @@
+                Ok(_) => continue,
+                Err(e) => error!("An error occurred while watching the plugin directory. Will continue to retry: {:?}", e),
+            }
+        }
+        Ok(())
+    }
+
+    async fn handle_create(&self, event: Event) -> anyhow::Result<()> {
+        // Filter paths, checking if it is a socket and not a directory. Why not check if it is a
+        // file? Because it isn't technically a regular file and so the `is_file` check returns
+        // false
+        for discovered_path in event
+            .paths
+            .into_iter()
+            .filter(|p| !p.is_dir() && p.extension().unwrap_or_default() == SOCKET_EXTENSION)
+        {
+            debug!(
+                "Beginning plugin registration for plugin discovered at {}",
+                discovered_path.display()
+            );
+
+            // Step 1: Attempt to call the socket. If this fails, we don't have any guarantee we'll
+            // be able to inform it we've failed. So just unwrap the error here
+            let plugin_info = get_plugin_info(&discovered_path).await?;
+            debug!(
+                "Successfully retrieved information for plugin discovered at {}:\n {:?}",
+                discovered_path.display(),
+                plugin_info
+            );
+
+            // Step 2: Validate discovered data
+            if let Err(e) = self.validate(&plugin_info, &discovered_path).await {
+                inform_plugin(&discovered_path, Some(e.to_string())).await?;
+                return Err(e).with_context(|| {
+                    format!(
+                        "Validation step failed for plugin discovered at {}",
+                        discovered_path.display()
+                    )
+                });
+            }
+            debug!(
+                "Successfully validated plugin discovered at {}:\n {:?}",
+                discovered_path.display(),
+                plugin_info
+            );
+
+            // Step 3: Register plugin to local storage
+            self.register(&plugin_info, &discovered_path).await;
+
+            // Step 4: Inform plugin
+            inform_plugin(&discovered_path, None).await?;
+            debug!("Plugin registration complete for {:?}", plugin_info)
+        }
+        Ok(())
+    }
+
+    async fn handle_delete(&self, event: Event) {
+        let mut lock = self.plugins.write().await;
+        for deleted_plugin in event
+            .paths
+            .into_iter()
+            .filter(|p| !p.is_dir() && p.extension().unwrap_or_default() == SOCKET_EXTENSION)
+        {
+            let key = match lock.iter().find(|(_, v)| *v.plugin_path == deleted_plugin) {
+                // Take ownership of the key to avoid an immutable borrow
+                Some((key, _)) => key.to_owned(),
+                // If for some reason it is already gone, no need to error
+                None => continue,
+            };
+            lock.remove(&key);
+        }
```

Comment:
Consider pulling this block out into a helper function `remove_plugin(plugins /* or lock */, deleted_plugin)`.  There is a lot of Rustish ceremony about `iter` and `find` and `to_owned` here that obscures the very simple intent.